src/protocol.ts

```diff
@@ -0,0 +1,313 @@
+import { Uint8ArrayList } from 'uint8arraylist'
+import { fromString as uint8ArrayFromString } from 'uint8arrays'
+import { alloc as uint8ArrayAlloc } from 'uint8arrays/alloc'
+import { InvalidCryptoExchangeError } from './errors.js'
+import { Nonce } from './nonce.js'
+import type { ICipherState, ISymmetricState, IHandshakeState, KeyPair, ICrypto } from './types.js'
+
+// Code in this file is a direct translation of a subset of the noise protocol https://noiseprotocol.org/noise.html,
+// agnostic to libp2p's usage of noise
+
+export const ZEROLEN = uint8ArrayAlloc(0)
+
+interface ICipherStateWithKey extends ICipherState {
+  k: Uint8Array
+}
+
+export class CipherState implements ICipherState {
+  public k?: Uint8Array
+  public n: Nonce
+  private readonly crypto: ICrypto
+
+  constructor (crypto: ICrypto, k: Uint8Array | undefined = undefined, n = 0) {
+    this.crypto = crypto
+    this.k = k
```

Comment:
Single letter variable names should at least have comments for clarity

Reply:
If you use an IDE, all protocol-related types have comments.